src/bootstrap/bootstrap.py

```diff
@@ -574,20 +586,59 @@ def download_toolchain(self):
                 rust_stamp.write(key)
 
     def _download_component_helper(
-        self, filename, pattern, tarball_suffix, rustc_cache,
+        self, key, filename, pattern, tarball_suffix,
     ):
-        key = self.stage0_compiler.date
-
+        cache_dst = os.path.join(self.build_dir, "cache")
+        rustc_cache = os.path.join(cache_dst)
         tarball = os.path.join(rustc_cache, filename)
         if not os.path.exists(tarball):
-            get(
-                self.download_url,
-                "dist/{}/{}".format(key, filename),
+            try:
+                get(
+                self.artifacts_server,
+                "{}/{}".format(key, filename),
                 tarball,
                 self.checksums_sha256,
                 verbose=self.verbose,
+                verify_checksum=False,
             )
+            except:
+                return False
         unpack(tarball, tarball_suffix, self.bin_root(), match=pattern, verbose=self.verbose)
+        return True
+
+
+
+    def download_bootstrap(self, commit_hash):
+        filename = "{}-{}.tar.xz".format("bootstrap-nightly", self.build_triple())
+        success = self._download_component_helper(commit_hash, filename, "bootstrap", "tar.xz")
+
+        return success
+
+    def last_bootstrap_commit(self):
+        top_level = output(["git", "rev-parse", "--show-toplevel"])
+
+        cmd = ["git", "rev-list", "--author={}".format(self.git_merge_commit_mail),
+               "-n1", "--first-parent", "HEAD"]
+        merge_base = output(cmd)
+
+        path = "{}/src/bootstrap".format(top_level)
+        cmd = ["git", "diff-index", "--quiet", "HEAD", "--", path]
+
+        ret = subprocess.Popen(cmd, cwd=self.rust_root)
+        code = ret.wait()
+        if code != 0:
+            return None
+        else:
+            return merge_base
+
+    def bootstrap_out_of_date(self, commit):
+        stamp_path = pathlib.Path(self.bin_root()).joinpath("bootstrap/.bootstrap-stamp")
+        if not stamp_path.exists():
+            return True
+
+        return stamp_path.read_text != commit
```

Comment:
This seems wrong as it will always return `True` because we don't have the stamp file on fresh repository.